### PR TITLE
Adjust time infrastructure

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes
For Cce, and especially for a combined Cce/GH evolution system, we need a bit more flexibility with the time stepper actions. This PR introduces the changes that I think will be sufficient to get those pieces working.

I'm open to alternative suggestions for how to better accomplish the same task. This version does have the unfortunate side-effect of introducing the empty tensor args in several existing executables.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by wthrowe** [comment]
Is there a reason you are avoiding Variables?  This is one of the things it was designed for.

**Comment by moxcodes** [comment]
My plan was to put the things I need to evolve in a Variables, but there's a couple of design challenges that could probably be worked around in different ways
- For most of the Cce hypersurface, the 'evolved' quantity and its corresponding dt counterpart are treated similarly and so far I've been putting them in the same Variables along with the other quantities that share similar properties. There's some inclination that having them all in the same memory block might help performance of spin-weighted derivatives, but I'm honestly not sure how much that gains us.
- I'll need to evolve both a real tensor quantity (the coordinates) and a complex scalar quantity (the Bondi evolved value), and our current architecture forbids things with different vector types being in the same variables.

So, I could probably get away with only the simpler of these two changes (just templating the existing utility) in exchange for making my Cce typelists somewhat more messy by ensuring that the evolved quantity and its time derivative have to be split out into two additional `Variables`

**Comment by wthrowe** [comment]
I'll ping @nilsdeppe for his thoughts on memory layout.

In the partial derivative code we do math on portions of Variables.  I don't know if that would be useful for the spatial derivatives.  I think the current time-stepping code is too high level to do that without some wrapping.

In any case, having both real and complex values pretty much rules out doing this with a single call to the time stepper.

Templating UpdateU, etc., makes sense.

**Comment by nilsdeppe** [comment]
Yes, the current partial derivatives code supports operating on a subset of a Variables. I will point out that it's done via pointer manipulation directly and not using a non-owning `Variables` (which I think we probably want at some point, actually).

I think in theory we could support Variables with mixed complex<double> and double, my only concern is that it would need to be tested extremely well because there'll be spooky memory things happening like reinterpret_cast.

Is that helpful?

**Comment by moxcodes** [comment]
I think the only reason to try to keep the swsh derivatives acting on a consistent Variables object is a rough reasoning with trying to keep caches warm when doing a bunch of transforms; the way we need to do it in different steps of the hypersurface computation, there's little hope of managing the transforms as true memory blocks, even if libsharp was optimized for it.

I wouldn't mind creating mixed variables, but that's going to be far harder than just templating the steppers.

Are we in agreement then that the best course here is to drop the tensor version from this PR and just work with templated `UpdateU` and `RecordTimeStepperData`?

**Comment by moxcodes** [comment]
This has been updated, dropping the tensor version of the generalization and just templating the existing `RecordTimeStepperData` and `UpdateU` to apply to general pairs of tags.

**Comment by kidder** [comment]
@wthrowe can you look at this?

File: tests/Unit/Time/Actions/Test_RecordTimeStepperData.cpp [hunk 1]
```diff
@@ -69,7 +106,15 @@ SPECTRE_TEST_CASE("Unit.Time.Actions.RecordTimeStepperData",
   history_tag::type history{};
   history.insert(TimeStepId(true, 0, slab.end()), 2., 3.);
 
+  alternative_history_tag::type alternative_history{};
+  alternative_history.insert(TimeStepId(true, 0, slab.end()), 2., 3.);
+
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<double> value_dist{1.0, 5.0};
```
Comment (wthrowe) [inline tests/Unit/Time/Actions/Test_RecordTimeStepperData.cpp]: These are unused.

**Review by wthrowe — CHANGES_REQUESTED** [review]
The design looks fine.

File: src/DataStructures/Tensor/Tensor.hpp [hunk 2]
```diff
@@ -622,6 +622,16 @@ struct MakeWithValueImpl<Tensor<double, Structure...>, double> {
   }
 };
 
+template <typename... Structure>
+struct MakeWithValueImpl<Tensor<DataVector, Structure...>, size_t> {
+  /// \brief Returns a Tensor whose elements are set equal to `value` of size
+  /// `size`
+  static SPECTRE_ALWAYS_INLINE Tensor<DataVector, Structure...> apply(
+      const size_t& size, const double value) noexcept {
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Tensor.hpp]: Can this just be `const size_t size` without a reference?

**Comment by moxcodes** [comment]
fixups posted.
Cheers!

**Comment by nilsdeppe** [comment]
LGTM! I'm happy for you to squash once @wthrowe has had a chance to look :)

**Comment by wthrowe** [comment]
Looks good.

**Comment by moxcodes** [comment]
Squashed.
Cheers!

**Review by nilsdeppe — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**